### PR TITLE
Histogram support for monitoring exporter

### PR DESCRIPTION
**Pull request by cnnradams**
`HistogramAggregator` is new with the Views API. Don't think this can be merged until 0.12.0 is released (or how does our releasing process work?)

### CODE REVIEW

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 1]
```diff
@@ -264,6 +283,7 @@ def export(
         for record in metric_records:
             instrument = record.instrument
             metric_descriptor = self._get_metric_descriptor(record)
+            print(metric_descriptor)
```
Comment (AndrewAXue) [inline opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:286]: Please remove

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 2]
```diff
@@ -280,16 +300,75 @@ def export(
...
+                        bucket_counts=[bucket[1] for bucket in buckets],
+                        bucket_options={
+                            "explicit_buckets": {
+                                # don't put in > bucket
+                                "bounds": [
+                                    bucket[0] for bucket in buckets[:-1]
+                                ]
+                            }
+                        },
+                        # exemplars=[
```
Comment (AndrewAXue) [inline opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:321]: please remove

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 3]
```diff
@@ -280,16 +300,75 @@ def export(
...
+            point_dict = {"interval": {}}
+
+            if isinstance(record.aggregator, HistogramAggregator):
+                count = sum(record.aggregator.checkpoint.values())
+                buckets = list(record.aggregator.checkpoint.items())
+
+                point_dict["value"] = {
+                    "distribution_value": Distribution(
+                        count=count,
+                        bucket_counts=[bucket[1] for bucket in buckets],
```
Comment (AndrewAXue) [inline opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:312]: Perhaps do the computation of bucket_counts and bucket bounds outside of here (maybe using a `zip`), its a little confusing for the reader who doesn't know what bucket[1] and bucket[0] represent.

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 4]
```diff
@@ -280,16 +300,75 @@ def export(
...
-                data_point = record.aggregator.checkpoint.last
+            point_dict = {"interval": {}}
+
+            if isinstance(record.aggregator, HistogramAggregator):
+                count = sum(record.aggregator.checkpoint.values())
+                buckets = list(record.aggregator.checkpoint.items())
+
+                point_dict["value"] = {
+                    "distribution_value": Distribution(
+                        count=count,
```
Comment (AndrewAXue) [inline opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:311]:
```suggestion
                buckets = list(record.aggregator.checkpoint.items())

                point_dict["value"] = {
                    "distribution_value": Distribution(
                        count=sum(record.aggregator.checkpoint.values()),
```

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 5]
```diff
@@ -280,16 +300,75 @@ def export(
...
+                    metric_descriptor.value_type
+                    == MetricDescriptor.ValueType.INT64
+                ):
+                    data_point = record.aggregator.checkpoint
+
+                elif (
+                    metric_descriptor.value_type
+                    == MetricDescriptor.ValueType.DOUBLE
+                ):
+                    data_point = record.aggregator.checkpoint
```
Comment (AndrewAXue) [inline opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:366]: Why is this necessary?

Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: idk, removed

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 4 above)
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: fixed in a better way

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 3 above)
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: done

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 2 above)
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:319]: added comment, I would prefer for this to stay unless it really needs to go

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 1 above)
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py]: done

File: opentelemetry-exporter-google-cloud/tests/test_cloud_monitoring.py [hunk 6]
```diff
@@ -414,6 +416,59 @@ def test_export_value_observer(self):
...
+                "display_name": "name",
+                "description": "description",
+                "labels": [],
+                "metric_kind": "CUMULATIVE",
+                "value_type": "DISTRIBUTION",
+            }
+        )
+
+        aggregator = HistogramAggregator(config={"bounds": [2, 4, 6]})
+        aggregator.checkpoint = OrderedDict([(2, 1), (4, 2), (6, 4), (">", 3)])
```
Comment (aabmass) [inline opentelemetry-exporter-google-cloud/tests/test_cloud_monitoring.py:444]: Could you fill me in on the `>` -- why is there no `<`?

File: opentelemetry-exporter-google-cloud/tests/test_cloud_monitoring.py [hunk 7]
```diff
@@ -414,6 +416,59 @@ def test_export_value_observer(self):
...
+        aggregator.last_update_timestamp = (WRITE_INTERVAL + 1) * int(1e9)
```
Comment (aabmass) [inline opentelemetry-exporter-google-cloud/tests/test_cloud_monitoring.py:445]: After rebase, you should be able to import `NANOS_PER_SECOND`

File: dev-constraints.txt [hunk 8]
```diff
@@ -10,5 +10,5 @@ Sphinx==3.1.2
 # development before GA. After GA, we will build against specific releases.
 # Bump the commit frequently during development whenever you are missing
 # changes from upstream.
--e git+https://github.com/open-telemetry/opentelemetry-python.git@8bc7786b191b4f1b1f161de983db71643bf9f51e#egg=opentelemetry-api&subdirectory=opentelemetry-api
--e git+https://github.com/open-telemetry/opentelemetry-python.git@8bc7786b191b4f1b1f161de983db71643bf9f51e#egg=opentelemetry-sdk&subdirectory=opentelemetry-sdk
+-e git+https://github.com/open-telemetry/opentelemetry-python.git@81025e1e159dc341d96a7c90df8f6e7b26e1cb90#egg=opentelemetry-api&subdirectory=opentelemetry-api
+-e git+https://github.com/open-telemetry/opentelemetry-python.git@81025e1e159dc341d96a7c90df8f6e7b26e1cb90#egg=opentelemetry-sdk&subdirectory=opentelemetry-sdk
```
Comment (aabmass) [inline dev-constraints.txt:14]: Keep these changes when you rebase, LGTM!

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 9]
```diff
@@ -280,16 +292,34 @@ def export(
                     UNIQUE_IDENTIFIER_KEY
                 ] = self.unique_identifier
 
-            point = series.points.add()
-            if isinstance(record.aggregator, SumAggregator):
-                data_point = record.aggregator.checkpoint
-            elif isinstance(record.aggregator, ValueObserverAggregator):
-                data_point = record.aggregator.checkpoint.last
+            point_dict = {"interval": {}}
```
[the remark this answers is not in the document]
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:295]: @AndrewAXue provided some base code for this and he used this method - maybe he can provide some insight?

File: opentelemetry-exporter-google-cloud/tests/test_cloud_monitoring.py (hunk 6 above)
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/tests/test_cloud_monitoring.py:444]: the current histogram aggregator defines buckets as everything < than bound[i] and >= than bound[i-1] - so you have a > for everything over the last bucket but your first bucket already covers everything below. This may change after the SDK spec is defined

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py [hunk 10]
```diff
@@ -186,17 +189,23 @@ def _get_metric_descriptor(
             descriptor["metric_kind"] = MetricDescriptor.MetricKind.CUMULATIVE
         elif isinstance(record.aggregator, ValueObserverAggregator):
             descriptor["metric_kind"] = MetricDescriptor.MetricKind.GAUGE
+        elif isinstance(record.aggregator, HistogramAggregator):
+            descriptor["metric_kind"] = MetricDescriptor.MetricKind.CUMULATIVE
```
[the remark this answers is not in the document]
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:193]: histogram is a monotonic counter for every bucket - so the metric kind is the same as counter.

File: opentelemetry-exporter-google-cloud/tests/test_cloud_monitoring.py (hunk 7 above)
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/tests/test_cloud_monitoring.py]: changed all 1e9 to NANOS_PER_SECOND

File: dev-constraints.txt (hunk 8 above)
Comment (cnnradams) [reply dev-constraints.txt:14]: Can you verify that I did it right?

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (AndrewAXue) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:295]: There was no real reason why it was originally done using a proto. I just wasn't aware you could even do it this way.

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (aabmass) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:193]: That makes sense. Is non-stateful Histogram supported in OT python and would that need to be Gauge?

[the remark this answers is not in the document]
Comment (cnnradams) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:193]: no, just like Counter, a non-stateful histogram is a cumulative but the start time is moved to the previous export time instead of always being at the start time of the program

File: opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py (hunk 9 above)
[the remark this answers is not in the document]
Comment (aabmass) [reply opentelemetry-exporter-google-cloud/src/opentelemetry/exporter/cloud_monitoring/__init__.py:295]: I prefer to use the the protos since they will surface validation errors where they are set. Created https://github.com/GoogleCloudPlatform/opentelemetry-operations-python/issues/56, lets go ahead with this for now

File: dev-constraints.txt (hunk 8 above)
Comment (aabmass) [reply dev-constraints.txt:14]: LGTM
